gemini/use-cases/retrieval-augmented-generation/qna_using_query_routing/utils/intent_routing.py

```diff
@@ -0,0 +1,413 @@
+        """
+
+        unable_to_understand_question = self.config["error_msg"][
+            "unable_to_understand_question"
+        ]
+        non_programming_question_error_msg = self.config["error_msg"][
+            "non_programming_question_error_msg"
+        ]
+        enabled_programming_language = self.config["default"][
+            "enabled_programming_language"
+        ]
+        default_language = self.config["default"]["default_language"]
+        chat = self.model.start_chat()
+        response = chat.send_message(
+            f"""
+        You are genai Programming Language Learning Assistant.
+        Your task is to undersand the question and write a code for same.
+
+        Instructions:
+        1. If programming language is not mentioned, then use {default_language} as default programming language to write a code.
+        2. Strictly follow the instructions mentioned in the question.
+        3. If the question is not clear then you can answer "{unable_to_understand_question}"
+        4. Strictly answer the question if only {enabled_programming_language} is mentioned in question.
+
+        If the question is about other programming language then DO NOT provide any answer, just say "{non_programming_question_error_msg}"
+
+        """
+        )
+
+        response = chat.send_message(
+            f"""{text}""", generation_config=self.genai_chat_parameters
+        )
+        if response.to_dict()["candidates"][0]["finish_reason"] != 1:
+            self.logger.info(
+                "ask_codey: No response from QnA due to LLM safety checks."
+            )
+            self.logger.info("LLM error code: %s\n", response.raw_prediction_response)
+        response = response.text
+        response = response.replace("```", "\n\n```")
+        response = response.replace("```java", "``` java")
+        return response
+
+    def get_programming_lanuage_from_query(
+        self, text: str, enabled_programming_language: List
+    ) -> List[str]:
+        """
+        Extracts programming languages mentioned in a user's query.
+
+        Args:
+            text (str): The user's query.
+            enabled_programming_language (List): List of supported languages.
+
+        Returns:
+            list: A list of programming languages extracted from the query (potentially empty).
+        """
+
+        response = self.model.generate_content(
+            f"""
+            You are strict programming languages extractor.
+
+            Instructions:
+            1. Extract only programming languages from message.
+            2. Don't return any other languages other than programming.
+            3. return [] if no programming lanuage in mentioned in message.
+            4. {enabled_programming_language} these are the programming languages.
+
+            Examples:
+            write a code for fibonacci series in C++? : ["C++"]
+            write a code using C# to generate palindrome series : ["C#"]
```

Comment:
```suggestion
            write code using C# to generate palindrome series : ["C#"]
```

Change this throughout. Most engineers would say "write code" instead of "write a code"

Reply:
updated the instruction in prompt.